### PR TITLE
[PM-12700] Add private key regeneration process

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-12700

## 📔 Objective

Requires server work in https://github.com/bitwarden/server/pull/4929

The purpose of this PR is to add the private key regeneration process on UI refreshed login and lock components.
Old components are out of scope for this feature.

This feature is behind a feature flag.

Target flows are: 

- Login
- Lock
- Log in with device
- Passkey Login (this currently is web client only and requires the passkey to be “used for encryption”)

Out of scope are the login flows: 

- SSO Login
- 2FA Login

Private key regeneration will be added to these once the UI refresh rewrite for them is merged into main in follow-up PRs. 

The `LoginSuccessHandlerService` was the requested approach for the auth team and why it was created under their ownership.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Comment by Thomas-Avery** [comment]
Looks like those tests are failing on main and aren't related to this PR.

**Review by mzieniukbw — APPROVED** [review]
LGTM. I have added few optional changes.

File: libs/key-management/src/user-asymmetric-key-regeneration/services/default-user-asymmetric-key-regeneration.service.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      );
+    }
+  }
+
+  private async shouldRegenerate(userId: UserId): Promise<boolean> {
+    const [userKey, userKeyEncryptedPrivateKey, publicKeyResponse] = await Promise.all([
+      firstValueFrom(this.keyService.userKey$(userId)),
+      firstValueFrom(this.keyService.userEncryptedPrivateKey$(userId)),
+      this.apiService.getUserPublicKey(userId),
+    ]);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply libs/key-management/src/user-asymmetric-key-regeneration/services/default-user-asymmetric-key-regeneration.service.ts]: I like it, updated in https://github.com/bitwarden/clients/pull/11829/commits/d677b9e9359d23f1806b82542d3e0bde0b6779ab

File: libs/auth/src/common/abstractions/login-success-handler.service.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { UserId } from "@bitwarden/common/types/guid";
+
+export abstract class LoginSuccessHandlerService {
+  /**
+   * Runs any service calls required after a successful login.
+   * Service calls that should be included in this method are only those required to be awaited after successful login.
+   * @param userId The user id.
+   */
+  abstract run: (userId: UserId) => Promise<void>;
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply libs/auth/src/common/abstractions/login-success-handler.service.ts]: I think this is what you're asking for ? https://github.com/bitwarden/clients/pull/11829/commits/5466be05bfd3339f4929839c84ac949e8d31ee90

It looks like KM follows a standard, so I swapped to that. In the rest of the code base, they're both used. 

Not sure what the difference is?

[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/auth/src/common/abstractions/login-success-handler.service.ts]: Methods declared as arrow functions in an abstract class are treated as properties in TypeScript, not as traditional methods. Generally, we would want actual functions to be treated as functions. It makes things like testing easier as well.

File: package-lock.json
Comment (JaredSnider-Bitwarden) [inline package-lock.json:1]: Should this have been updated here?

File: package.json
Comment (JaredSnider-Bitwarden) [inline package.json:1]: Should this have been updated in this PR?

File: package-lock.json
Comment (Thomas-Avery) [reply package-lock.json:1]: Yes, we need the newer version of the SDK that has the private key regeneration methods.

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Auth files look great! Tyvm for implementing the new service we discussed!

**Review by mzieniukbw — APPROVED** [review]
LGTM

**Pull request merged**